builtins/web/fetch/headers.cpp

```diff
@@ -555,24 +543,28 @@ JSObject *Headers::create(JSContext *cx, HandleValue init_headers, host_api::Htt
   if (!self) {
     return nullptr;
   }
-  return init_entries(cx, self, init_headers);
+  if (!init_entries(cx, self, init_headers)) {
+    return nullptr;
+  }
+  MOZ_ASSERT(mode(self) == Headers::Mode::ContentOnly || mode(self) == Headers::Mode::Uninitialized);
```

Comment:
Right now, I'm pretty sure `mode(self)` will always be `Unitialized` here, because `init_entries` doesn't change it. And I think here we need to do the same thing as for the other `create` overload and explicitly set the state. Otherwise, `switch_mode` will reset the `entries` map to an empty `Map` at the next opportunity.

Which I guess in turn means that we can simplify `switch_mode` even further, because we should now have a setup in which it can never encounter the `Unitialized` state.
```suggestion
  MOZ_ASSERT(Headers::mode(self) == Headers::Mode::Uninitialized);
  SetReservedSlot(self, static_cast<uint32_t>(Headers::Slots::Mode),
                  JS::Int32Value(static_cast<int32_t>(Headers::Mode::ContentOnly)));
```

Reply:
In this variation of headers create, `headersInit` contains iterable header values to populate the initial headers. So when `init_entries` does that population, the state is changed to `ContentOnly` (through append_value calling prepare_entries_for_modification), unless the `headersInit` was empty in which case the headers remain in `Uninitialized`.

So the assertion remains correct and non-redundant.

Reply:
Oh, of course! This only goes to show that simplifying this code is a very good idea! 😬